Feature/intersection tables

jacob6838 opened the pull request:
<!--- Thanks for the contribution! -->

# PR Details

## Description

**DISCLAIMER** This is not intended to be included in the current JPO-USDOT release. As such, this has been created as a Draft PR.

This PR adds the management of intersection data to the cvmanager. This includes the following:
1. Adding postgresql tables for intersections, intersection_organizations, and rsu_intersections
2. Adding API endpoints to handle management of intersections and links
3. Adding webapp Admin pages for creating, editing, and removing intersections, intersection-organization links, and rsu-intersection links.

## How Has This Been Tested?

These changes were tested locally, running the api, webapp, and database updates together. Through the UI, you are able to create intersections, remove intersections, edit intersections, create/remove intersection - organization links, and create/remove rsu - intersection links. 

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] My changes require new environment variables:
  - [ ] I have updated the docker-compose, K8s YAML, and all dependent deployment configuration files.
- [ ] My changes require updates to the documentation:
  - [ ] I have updated the documentation accordingly.
- [x] My changes require updates and/or additions to the unit tests:
  - [x] I have modified/added tests to cover my changes.
- [x] All existing tests pass.

jacob6838 committed:
Adding intersection tables to create sql

jacob6838 committed:
Adding intersection admin pages

jacob6838 committed:
Final intersection admin logical touches

jacob6838 committed:
Merge branch 'develop' into Feature/intersection-tables

jacob6838 committed:
Updating intersection unit tests

dmccoystephenson commented on resources/sql_scripts/CVManager_CreateTables.sql:
```diff
@@ -308,6 +308,7 @@ CREATE TABLE IF NOT EXISTS public.rsu_organization
       ON DELETE NO ACTION
 );
 
+-- TODO: should combine with intersection table?
```
Should this TODO be taken care of in this PR? If not, should we open a work item for this?

jacob6838 replied:
Replaced this with a new work item: https://dev.azure.com/SOC-OIT/CDOT/_workitems/edit/263317

dmccoystephenson reviewed (changes requested):
The API unit tests have some failures:

![image](https://github.com/user-attachments/assets/e4bdc5a5-c7f7-43e5-9b0d-5e8a8acdf9ba)

jacob6838 committed:
Merge branch 'develop' into Feature/intersection-tables

jacob6838 committed:
Updating intersection unit tests

jacob6838 committed:
Update CVManager_CreateTables.sql

dmccoystephenson reviewed (approved):
This looks good to me! The API/webapp unit tests pass & I am able to access the new UI pages through the webapp.

jacob6838 committed:
Merge branch 'develop' into Feature/intersection-tables

drewjj commented on resources/sql_scripts/CVManager_CreateTables.sql:
```diff
@@ -459,4 +459,70 @@ CREATE TABLE IF NOT EXISTS public.obu_ota_requests (
 
-CREATE SCHEMA IF NOT EXISTS keycloak;
\ No newline at end of file
+CREATE SCHEMA IF NOT EXISTS keycloak;
```
Can you add a SQL script with just these new table CREATEs in the update_scripts directory? Will be handy for making the updates without much thought.

jacob6838 replied:
added!

drewjj commented on services/api/src/admin_org.py:
```diff
@@ -270,6 +329,20 @@ def check_orphan_rsus(org):
     return False
 
 
+def check_orphan_intersections(org):
```
Nice, glad this is being accounted for.

drewjj commented on webapp/src/features/adminEditIntersection/AdminEditIntersection.tsx:
```diff
@@ -0,0 +1,334 @@
+    setValue,
+  } = useForm<AdminEditIntersectionFormType>({
+    defaultValues: {
+      intersection_id: '',
```
Is there much value to allowing intersection ID to be seen on the modification when the API doesn't actually allow for this to be modified?

jacob6838 replied:
The intersection_id corresponds to the intersection_number in the sql table. The intersection_id (intersection_number in sql) I believe should be able to be changed

jacob6838 replied:
Added the ability to modify the intersection ID

drewjj commented on webapp/src/features/adminAddIntersection/AdminAddIntersection.tsx:
```diff
@@ -0,0 +1,219 @@
+          style={{ fontFamily: 'Arial, Helvetica, sans-serif' }}
+        >
+          <Form.Group className="mb-3" controlId="ip">
+            <Form.Label>Intersection ID</Form.Label>
```
Is the intersection ID necessary to allow the user to define it? Is this the value that gets used to match corresponding MAP and SPaT messages to?

jacob6838 replied:
This is the intersection_number in the database, used to link up MAP and SPAT messages

drewjj commented on services/api/tests/src/test_admin_org.py:
```diff
@@ -282,20 +286,52 @@ def test_delete_org(mock_query_db, mock_write_db):
 
+
 @patch("api.src.admin_org.pgquery.query_db")
 def test_delete_org_failure_orphan_rsu(mock_query_db):
```
While the API unit tests pass, we seem to be missing tests for all of the new python scripts you've added. We will want to  add some tests for these.

jacob6838 replied:
Added and passing

drewjj reviewed (changes requested):
Everything was put together nicely and the unit tests pass for both the webapp and api. A number of comments I would like addressed but it is very close!

jacob6838 committed:
Creating intersection api unit tests

jacob6838 committed:
Finalizing intersection api unit tests

jacob6838 committed:
Create intersection_update.sql

jacob6838 committed:
enabling editing of intersection ID

jacob6838 committed:
Update admin_intersection.py

jacob6838 committed:
Update intersection_update.sql

drewjj reviewed (approved):
Nice, everything looks great!

payneBrandon commented on services/api/tests/src/test_admin_org.py:
```diff
@@ -217,6 +219,8 @@ def test_modify_user_success(mock_pgquery, mock_check_safe_input):
         call(admin_org_data.modify_org_remove_user_sql),
         call(admin_org_data.modify_org_add_rsu_sql),
         call(admin_org_data.modify_org_remove_rsu_sql),
+        call(admin_org_data.modify_org_add_intersection_sql),
```
Just noticing with these additions, but we're calling admin_org.modify_org and the test method is called "test_modify_user_success". Should we rename to reflect what we're really doing here?

jacob6838 replied:
re-named to "test_modify_organization_success"

payneBrandon commented on services/api/tests/src/test_admin_new_intersection.py:
```diff
@@ -0,0 +1,113 @@
+        assert body == {}
+
+
+def test_entry_post_schema():
```
Are we testing a bad schema here? Should we name it to reflect?

jacob6838 replied:
renamed to "test_entry_post_schema_bad_json"

payneBrandon commented on services/api/tests/src/test_admin_intersection.py:
```diff
@@ -0,0 +1,285 @@
+from unittest.mock import patch, MagicMock, call
```
This file seems to be predominantly copy/paste from test_admin_rsu.py, to the point we're even importing the admin_rsu and admin_rsu_data.

jacob6838 replied:
Good catch, went though and updated all of these tests

payneBrandon reviewed (changes requested):
Service functionality looks good, but I have a few questions with the unit tests

payneBrandon commented on webapp/src/features/adminAddIntersection/adminAddIntersectionSlice.tsx:
```diff
@@ -0,0 +1,256 @@
+  submitAttempt: false,
+}
+
+export const updateApiJson = (apiJson: AdminIntersectionCreationInfo): AdminIntersectionKeyedCreationInfo => {
```
Can we add some descriptions to the methods? "UpdateApiJson" is fairly ambiguous as to what the method does.

jacob6838 replied:
Updated the name to "convertApiJsonToKeyedFormat"

jacob6838 committed:
Fixing intersesction unit tests, renaming updateApiJson

jacob6838 committed:
Update test_admin_org.py

jacob6838 committed:
Merge branch 'develop' into Feature/intersection-tables

payneBrandon commented on webapp/src/features/adminAddIntersection/adminAddIntersectionSlice.tsx:
```diff
@@ -0,0 +1,258 @@
+  }
+}
+
+export const updateJson = (
```
We updated the other json conversion function name, but there's still some in here like this updateJson that we should rename and add descriptions to accordingly.

jacob6838 replied:
Added docstring comments and renamed method

payneBrandon commented on webapp/src/features/adminEditIntersection/adminEditIntersectionSlice.tsx:
```diff
@@ -0,0 +1,234 @@
+  return obj && Object.keys(obj).length === 0
+}
+
+export const checkForm = (state: RootState['adminEditIntersection']) => {
```
Similar to previous comments, let's add some commenting into these methods to know what they are doing. "checkForm" seems to only be validating a property length is not 0, "updateJson" is removing null properties and updating some lists but without digging into the code we'd have no idea that's what it does.

jacob6838 replied:
Added docstring comments and renamed method

payneBrandon commented on webapp/src/features/adminEditIntersection/adminEditIntersectionSlice.tsx:
```diff
@@ -0,0 +1,234 @@
+        return { name: val }
+      })
+      state.value.rsus = allowedSelections.rsus.map((val) => {
+        return { name: val?.replace('/32', '') }
```
Are we just updating the rsu ip address to be more human readable here by removing the CIDR subnet mask? Can we add some comments into these to say why we're doing some of this so we don't forget later?

jacob6838 replied:
Added comments, these are both for human readability and api compatibility (for the editIntersection API)

jacob6838 committed:
Adding docstring comments, renaming methods

payneBrandon commented:
Overall code is looking good, but we need some more descriptions in the code, better naming conventions, or both to help the developers understand what we're looking at for methods and why things are doing what they are.

jacob6838 committed:
Additional docstring comments and renaming

jacob6838 commented:
> Overall code is looking good, but we need some more descriptions in the code, better naming conventions, or both to help the developers understand what we're looking at for methods and why things are doing what they are.

I just went though and added docstring comments to all of the slice methods, and renamed the ambiguous methods

jacob6838 committed:
Update AdminIntersectionTab.tsx

jacob6838 committed:
Fixing intersection query for multiple RSU-Intersection links

payneBrandon reviewed (approved):
lgtm!
